TSVB add markdown tests

vitalics opened the pull request:
## Summary
add new tsvb markdown tests
Summarize your PR. If it involves visual changes include a screenshot or gif.

### Checklist

Use ~~strikethroughs~~ to remove checklist items you don't feel are applicable to this PR.

- [ ] This was checked for cross-browser compatibility, [including a check against IE11](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility)
- [X] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)
- [ ] [Documentation](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#writing-documentation) was added for features that require explanation or tutorials
- [X] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios
- [ ] This was checked for [keyboard-only and screenreader accessibility](https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility#Accessibility_testing_checklist)

### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)
- [ ] This includes a feature addition or change that requires a release note and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)

vitalics committed:
reorganize tsvb markdown test
add test for table checking

vitalics committed:
reorganize tsvb markdown test
add test for table checking

vitalics committed:
add new tests for markdown

vitalics committed:
add new tests for markdown

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/12256/)

elasticmachine commented:
Pinging @elastic/kibana-app

dmlemeshko commented on test/functional/apps/visualize/_tsvb_markdown.js:
```diff
@@ -0,0 +1,85 @@
+
+      it('should be show empty table variables', async () => {
+        const variables = await visualBuilder.getMarkdownTableVariables();
+        expect(variables).to.be('No variables available for the selected data metrics.');
```
`getMarkdownTableVariables` should do what it states: return list of `{name, value}`. If its empty => return `[]`
In case of checking message, use other function e.g. `getMarkdownTableMessage()`

dmlemeshko commented on test/functional/apps/visualize/_tsvb_markdown.js:
```diff
@@ -0,0 +1,85 @@
+      it('should be render all markdown components', async () => {
+        const table = await find.byCssSelector('table.table');
+        const isExisting = find.exists(table);
+        const tabsCount = await visualBuilder.getSubTabsCount();
```
1. The general rule for functional tests: use `testSubjects` over `find` service as much as possible.
In particular case, you need to add `data-test-subj="tsvbMarkdownVariablesTable"` for table here:
https://github.com/elastic/kibana/blob/master/src/legacy/core_plugins/metrics/public/components/markdown_editor.js#L172
Then you can use `testSubjects.find('tsvbMarkdownVariablesTable')`

2. Verifying table is present is **not good enough**: it can be empty as screenshot displays
<img width="1336" alt="tsvb_markdown" src="https://user-images.githubusercontent.com/10977896/52630756-633ab800-2ebd-11e9-81e6-23ae44b199d8.png">
This test should check variables list is rendered as it was frequent issue in the past

3. Avoid direct usage of `testSubject`/`find` services in test. Should be wrapped up in PageObject:
```
const variablesList = visualBuilder.getMarkdownTableVariables()
expect(variablesList).to.have.length(5)
```

dmlemeshko commented on test/functional/apps/visualize/_tsvb_markdown.js:
```diff
@@ -0,0 +1,85 @@
+          `{{#each _all}}
+- {{ data.formatted.[0] }} {{ data.raw.[0] }}
+{{/each}}`;
+
```
Fix formatting, you may use `\n`?

dmlemeshko replied:
```
const list = '{{#each _all}}\n{{ data.formatted.[0] }} {{ data.raw.[0] }}\n{{/each}}';
```

vitalics replied:
I missed this review 😢

dmlemeshko commented on test/functional/page_objects/visual_builder_page.js:
```diff
@@ -82,6 +83,64 @@ export function VisualBuilderPageProvider({ getService, getPageObjects }) {
+     * getting all markdown variables list which located on `table` section
+     *
+     * **Note**: if `table` not have variables, return `No variables available for the selected data metrics.`
+     * @returns {Promise<string[] | string>}
```
It is better to return `[{name, value}]`
Don't try to implement universal function, for error msg use another one :)

dmlemeshko commented on test/functional/page_objects/visual_builder_page.js:
```diff
@@ -82,6 +83,64 @@ export function VisualBuilderPageProvider({ getService, getPageObjects }) {
+        if (tab.getVisibleText() === subTab) {
+          tab.click();
+        }
+      });
```
One of the tabs has `data-test-subj="markdownDataBtn"` attribute. You need to add similar ones for other `EuiTab` elements and use it for elements allocation via `testSubject` service

dmlemeshko commented on test/functional/apps/visualize/_tsvb_markdown.js:
```diff
@@ -0,0 +1,85 @@
+        await timePicker.setAbsoluteRange('2015-09-22 06:00:00.000', '2015-09-22 11:00:00.000');
+      });
+
+      it('should be render all markdown components', async () => {
```
`should render all markdown components`

dmlemeshko commented on test/functional/page_objects/visual_builder_page.js:
```diff
@@ -82,6 +83,64 @@ export function VisualBuilderPageProvider({ getService, getPageObjects }) {
+      const elements = await rootSelector.findAllByCssSelector(tabsSelector);
+
+      return elements.length;
+    }
```
with `data-test-subj` in each of the button, you can simply use `[data-test-subj$="TsvbTypeBtn"]`:

`return await find.findAllByCssSelector ('[data-test-subj$="TsvbTypeBtn"]')`
I would define function as `getTSVBTabs` and return list of ElementWrapper objects

vitalics commented:
this tests related for #23418, #28168 and #30625

vitalics committed:
fix PR review issues

vitalics committed:
fix PR review issues

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/283/)

dmlemeshko commented on test/functional/apps/visualize/_tsvb_markdown.js:
```diff
@@ -0,0 +1,86 @@
+      it('should show empty table variables', async () => {
+        const variables = await visualBuilder.getMarkdownTableVariables();
+        expect(Array.isArray(variables)).to.be.ok();
+        expect(variables).to.be.empty();
```
Example: https://github.com/elastic/kibana/blob/master/test/api_integration/apis/management/saved_objects/relationships.js#L80

dmlemeshko commented on test/functional/page_objects/visual_builder_page.js:
```diff
@@ -78,8 +79,60 @@ export function VisualBuilderPageProvider({ getService, getPageObjects }) {
+      const variables = testTableVariables.findAllByCssSelector(variablesSelector);
+      const variablesText = Promise.all(variables.map(variable => variable.getVisibleText()));
+      log.debug(`markdown table variables is: ${variablesText}`);
+      return variablesText;
```
The idea is to get pairs {name, value}, it maybe easier to do this way:
```
const $ = await testTableVariables.parseDomContent();
$('tbody tr').toArray().map(tr => {
const variableData = $(tr).find('td).text().toArray();
return {name: variableData[0], value: variableData[1]}
});
```

dmlemeshko commented:
I added several more comments. Branch is ~120 commits behind master, suggesting to merge latest master.

vitalics committed:
remove unused services
fix flacky for markdown table variables

vitalics committed:
remove unused services
fix flacky for markdown table variables

dmlemeshko commented on test/functional/page_objects/visual_builder_page.js:
```diff
@@ -59,6 +60,17 @@ export function VisualBuilderPageProvider({ getService, getPageObjects }) {
       const input = await find.byCssSelector('.tvbMarkdownEditor__editor textarea');
+      await this.clearMarkdown();
+      await input.type(markdown);
+      await PageObjects.common.sleep(500);
```
I don't think we need `sleep` here. `waitForRenderingCount` for checking condition is doing `retry` up to ~40 sec

dmlemeshko commented on test/functional/page_objects/visual_builder_page.js:
```diff
@@ -69,23 +81,81 @@ export function VisualBuilderPageProvider({ getService, getPageObjects }) {
 
     async getMarkdownText() {
       const el = await find.byCssSelector('.tvbEditorVisualization');
+      await PageObjects.common.sleep(300);
```
Is there a way to replace it with any `wait condition`?

vitalics replied:
no, since getting element so fast and best option is use sleep with a timeout. Actual element text is 
![image](https://user-images.githubusercontent.com/8816260/52961335-bf18ab80-33ab-11e9-953f-490445b2d493.png) 

and current text for same element is 
![image](https://user-images.githubusercontent.com/8816260/52961364-d0fa4e80-33ab-11e9-9d0d-53357cc23666.png)

dmlemeshko commented on test/functional/page_objects/visual_builder_page.js:
```diff
@@ -94,6 +164,7 @@ export function VisualBuilderPageProvider({ getService, getPageObjects }) {
       await PageObjects.header.waitUntilLoadingHasFinished();
     }
 
+
```
Redundant empty line

dmlemeshko commented on test/functional/page_objects/visual_builder_page.js:
```diff
@@ -69,23 +81,81 @@ export function VisualBuilderPageProvider({ getService, getPageObjects }) {
+        const value = await subVars[1].getVisibleText();
+        return { key, value, selector };
+      }));
+      log.debug(`markdown table variables is: ${variablesKeyValueSelectorMap}`);
```
`markdown variables table is` or `markdown table variables are`

dmlemeshko reviewed (changes requested):
All good, but need to get rid of `sleep`: we are trying to remove existing ones, so lets not add new ones :)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1173/)

vitalics committed:
fix flaky tests

vitalics committed:
fix flaky tests

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1206/)

dmlemeshko commented:
@vitalics you need to merge the latest master to your branch.

vitalics commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1477/)

vitalics commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1528/)

vitalics commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1748/)

vitalics committed:
remove markdown tests from _tsvb_chart file

vitalics committed:
remove markdown tests from _tsvb_chart file

loixlab committed:
[Rollup] Add unit tests for Job listing page (#31689)

sqren commented:
@vitalics Is it on purpose that you requested review from APM UI?

vitalics commented:
sorry, I merged with master, but do this without rebase

sqren commented:
@vitalics No worries. Can you remove the teams you don't need feedback from?

vitalics commented:
Unfourchantly I can't. Could you please remove `canvas`, `ml-ui`, `security`, `design` and `apm-ui` team from reviewers

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2006/)

walterra committed:
[ML] Deprecate stateFactory(). (#31692)

This deprecates our own implementation of `stateFactory()`. This utility was able to manage shared state across angular and react code including persisting it to the URL using `AppState`. This PR replaces it with a simpler way using RxJs' Observables.

simianhacker committed:
[Logs UI] Add link-to router for logs (without a node) (#31653)

* [Logs UI] Add link-to router for logs (without a node)

* Fixing snapshots

* Update x-pack/plugins/infra/public/pages/link_to/redirect_to_logs.tsx

Co-Authored-By: simianhacker <chris@chriscowan.us>

joshdover committed:
Add RFC README and template (#31650)

w33ble committed:
chore: unify the workpad update handler (#31218)

* test: assets and structures routes

* chore: unify the workpad update handler

the 3 functions were pretty much the same already

nreese committed:
Update embeddables metadata to support multiple index patterns (#31538)

* Update embeddables metadata to support multiple index patterns

* ts lint errors

* another try at ts lint errors

* fix dashboard_state tests

* use lodash compact to remove undefined

* revert _.compact to make TS linter happy

* rename simulateNewEmbeddableWithIndexPattern => simulateNewEmbeddableWithIndexPatterns

tylersmalley committed:
[UA] Protects against Cloud plugin being disabled (#31637)

Currently, if the Cloud plugin is disabled the server will fail to start

Fixes #31612

Signed-off-by: Tyler Smalley <tyler.smalley@elastic.co>

joshdover committed:
Add new platform code to dev watcher (#31622)

cjcenizal committed:
Fix Rollup Job List test, which broke due to change from jobs prop to hasJobs (#31830)

cqliu1 committed:
Refactor time_filter (#31641)

Fixes time_filter state

Removed console logs

Cleaned up changes

Removed unused prop in time_filter

Moved propTypes

Fixed bug with changing columns

Added comment

nreese committed:
[Maps] fix app crash where layer details panel calls getId on undefined (#31816)

* [Maps] fix app crash where details panel calls getId on undefined

* add jest test ensuring no errors when selectedLayer is undefined

sorenlouv committed:
[APM] Avoid crashing transaction details page if trace duration is 0 (#31799)

* [APM] Avoid crashing transaction details page if trace duration is 0

* Remove snapshot

tylersmalley committed:
[UA] Support previously re-indexed ML/Watcher indices (#31046)

* [UA] Support previously re-indexed ML/Watcher indices

Previously we were comparing the start of the indices to determine if ML/Watcher
should be stopped before re-indexing. Since we are now pre-pending the index
name we should compare the index names without the re-indexed portion.

Signed-off-by: Tyler Smalley <tyler.smalley@elastic.co>

w33ble committed:
fix: hours label container size (#31838)

tylersmalley committed:
[logging] Corrects intercept of ECONNRESET (#31742)

errno is not set for ECONNRESET

https://github.com/hapijs/wreck/blob/29d8f4b5eaa12c5d44c642130cd6002c1d83ffd2/lib/index.js#L290

Closes https://github.com/elastic/kibana/issues/22225

Example:

{ Error: socket hang up
  at TLSSocket.onSocketClose (_tls_wrap.js:761:23)
  at TLSSocket.emit (events.js:194:15)
  at _handle.close (net.js:600:12)
  at Socket.done (_tls_wrap.js:388:7)
  at Object.onceWrapper (events.js:277:13)
  at Socket.emit (events.js:189:13)
  at TCP._handle.close (net.js:600:12) code: 'ECONNRESET'

Signed-off-by: Tyler Smalley <tyler.smalley@elastic.co>

cjcenizal committed:
Fix spacing error in license header in CCR, Remote Clusters, and Rollups. (#31869)

* Remove duplicate license header.

vitalics committed:
skip first table test for ci check

vitalics committed:
rebase

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/2232/)

vitalics commented:
reopened #31893
